design-decisions.md

```diff
@@ -0,0 +1,28 @@
+# Design Decisions
+Assumptions, thoughts and decisions for current and future consideration.
+
+##  Technologies and frameworks
+- Typescript
+- TSLint
+- DynamoDb - The problem prefers it and it has good options for local development in a docker image or serverless plugin
+- Documentation - Rely on github for now but it would be nice to have polished api documentation
+- Serverless provisioning (serverless, AWS Sam, Terraform...) - I'd like to try out AWS Sam but I'm familiar with serverless so I'll choose it for this project
+- Local dev environment - I would normally begin with a fully functioning local dev environment (serverless-offline, serverless-dynamodb-local) - This is going to take a little more time but I still think it's essential to start any real project this way.  I'll invest the extra time on this.
+- serverless-dynamodb-local - We have other options with dynamodb-local through [docker](https://hub.docker.com/r/amazon/dynamodb-local) which I've used recently.  But I really wanted to evaluate this library. Unfortunately it requires Java.  For me that's not a huge issue but something that could become an issue for larger teams.
+
+- Validation - Try to handle with Json Shema and API Gateway.  I was able to get validation working with API Gateway and a schema defined in serverless BUT there is an open issue in serverless-offline which means it will not work locally.  Thus I wasn't able to add tests for it.  See below for a more optimal long-term approach.  
+- Testing (mocha, tsnode, jest) - I chose jest with ts-jest.  It worked well enough.  I ran into an issue where my unit tests really weren't going to offer much value but integration tests would. Integration tests working with all the local tooling to replicate the aws serverless environment with a single command would certainly be possible and optimal but was taking too much time.  For now I will run two commands to run tests :(
```

Comment:
> How did you come to that conclusion? How do you tend to evaluate testing needs for your code?
@afaq-sf Good question.  Lets look at the opportunities for unit testing here.
- `UserController.ts`  I'm looking for "units" where some value would be gained by testing them individually.  At first there was some value to testing this.  But it eventually became mostly a pure delegator with little or no actual business logic.  It's value at all can be debated separately.  At that point I have to mock the data access layer.  Then I'm writing tests like
```
expect(mockedDataAccess.createUser).toBeCalledWith(exactUserCreateObject);
```
This offers very little value.  Confirming that I've passed objects through unmodified.  Mocking objects is not always trivial.  It could have some very small value to prevent regressions but it's value is unlikely to ever be realized against the cost of maintaining these tests through refactors and enahancements.  

With an integration test I get all the coverage with much less cost.  This is a balancing act.  

Another example is
- ` ExceptionMapper.ts`  
I get 100% coverage for this class in my integration tests but this is an example where Unit Tests would have very little cost.  Also it may not be obvious as I'm changing my integration tests that I'm expected to cover this class.  `ExceptionMapper` is a good example, especially as it gets more complicated, of a class we should be unit testing.  It will be easy to prevent regressions and enhance this class.  

- `UserDataAccess`:  This class gets it's coverage from the integration tests.  Unit testing will have the overhead of mocking the DynamoDb DocumentClient.  This isn't trivial but this class does actually contain some business logic.  We could test that it errors when expected.  We could test that id's and audit columns (updated/created) are populated when expected.  I've found that these types of tests have value early on in a project and then you begin the repeat the same functionality over and over.  I usually add them early and decide potentially to remove them later when it makes sense and you have the coverage from other tests.

Tests are extremely valuable.  They allow us to move faster with confidence.  Writing testable code inherently produces more maintainable and expandable code.  However, every line of code you add has a cost in the short and long term.  You need to balance these investments